### ISSUE TITLE
setFieldValue deletes the value from values and prevents errors from showing on submit

### ISSUE THREAD

**Issue by amir-virta** [jaredpalmer/formik#2332]
### Current Behavior

When setFieldValue() sets a field to undefined it deletes it from `values`. The consequence of this is that Formik no longer marks the field as touched on submit and therefore error messages for the field no longer show (using `<ErrorMessage>`)

Steps to repro:
1) Formik state after submit is clicked and input field is not touched. As you can see the error message is visible.

![image](https://user-images.githubusercontent.com/38225887/75621864-cf32a180-5b4e-11ea-9c29-33c45723f6d7.png)

2) Formik state after (a) input field is set to a value, then (b) the input field is cleared, and then (c) submit is clicked again. The nuance here is that when the field is cleared, `setFieldValue` is called to set the form field to undefined. You can see the error exists, but the field name is no longer in values and touched does not have the field name either.

![image](https://user-images.githubusercontent.com/38225887/75621868-d8237300-5b4e-11ea-88ad-ea418632c222.png)

### Expected behavior

On submit, Formik should continue to mark the fields set to undefined as touched. 

### Reproducible example

https://codesandbox.io/s/formik-setfieldvalue-deletes-the-value-from-values-and-prevents-errors-from-showing-on-submit-ysxdh?fontsize=14&hidenavigation=1&theme=dark

### Suggested solution(s)

Two ways of this happening (a) set the field value to undefined instead of deleting it from values when `setFieldValue` is called with the value of undefined OR (b) clone initialValues and use that to determine which fields to touch on submit instead of the current set of keys in `values`.

### Additional context

None

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           |  2.1.1 (and 1.1.1) |
| React            | 16.12.0 |
| TypeScript       |  n/a |
| Browser          | Chrome Version 80.0.3987.122 (Official Build) (64-bit) |
| npm/Yarn         | Yarn |
| Operating System | Mac OS X |

**Comment by justinbhopper** [comment]
I agree, formik should be setting the value to undefined, not deleting the property from the values object.  Many operations in Formik rely on the existence of the entry in the values object in order to know what is errored/touched, and by deleting, it is wiping that data.

These lines seem to be the culprits, although there may be others:
https://github.com/jaredpalmer/formik/blob/master/packages/formik/src/utils.ts#L131
https://github.com/jaredpalmer/formik/blob/master/packages/formik/src/utils.ts#L139

**Comment by clausreinke** [comment]
Those two lines lead to #781, which attempts to fix #727. The PR title suggests this is intended behaviour for `setIn`, but it seems the original issue was only about `setFieldError`, where it makes sense for the keys to be removed.

Was it really the intention to have `setFieldValue` be affected by this?

**Comment by circAssimilate** [comment]
Thanks for filing this, I suspected that this "feature" we were leaning on was actually a bug! Fwiw: I created https://github.com/formik/formik/issues/2598 to backfill this functionality.

**Comment by MatejFacko** [comment]
Guys, any update on this? Seems to me that the issue is still there.

**Comment by justinbhopper** [comment]
For anyone who has run into this issue and needs a workaround, one solution is to create your own wrapper around Formik's setFieldValue function to override its behavior.  You can do this by either wrapping Formik entirely (which is more work but allows for more seamless integration) or just calling a surrogate function manually.

Here is what a surrogate function would look like:
```
function setFieldValue(formik: Formik<Values>, field: string, value: any, shouldValidate?: boolean): void {
  if (value === undefined) {
    const values = setIn(formik.state.values, field, undefined);
    formik.setValues(values);

    const willValidate = shouldValidate === undefined ? this.props.validateOnChange : shouldValidate;
    if (willValidate) {
      formik.validateForm(values);
    }
  } else {
    formik.setFieldValue(field, value, shouldValidate);
  }
}
```

**Comment by zyofeng** [comment]
> For anyone who has run into this issue and needs a workaround, one solution is to create your own wrapper around Formik's setFieldValue function to override its behavior. You can do this by either wrapping Formik entirely (which is more work but allows for more seamless integration) or just calling a surrogate function manually.
> 
> Here is what a surrogate function would look like:
> 
> ```ts
> function setFieldValue(formik: Formik<Values>, field: string, value: any, shouldValidate?: boolean): void {
>   // Override default behavior by forcing undefined to be set on the state
>   if (value === undefined) {
>     const values = setIn(formik.state.values, field, undefined);
>     formik.setValues(values);
> 
>     const willValidate = shouldValidate === undefined ? this.props.validateOnChange : shouldValidate;
>     if (willValidate) {
>       formik.validateForm(values);
>     }
>   } else {
>     // Use default behavior for normal values
>     formik.setFieldValue(field, value, shouldValidate);
>   }
> }
> ```

This doesn't work as setIn method is the one that removes the obj if value is undefined. see line 130

https://github.com/formium/formik/blob/2d613c11a67b1c1f5189e21b8d61a9dd8a2d0a2e/packages/formik/src/utils.ts

**Comment by justinbhopper** [comment]
@zyofeng You're right, my apologies.  I failed to paste the important part, which was a replacement of setIn that I was using that changes the default setIn behavior.

**Comment by Kepro** [comment]
This is an annoying issue, I spent an hour figuring out the issue, and it's just because if a value is `undefined`, Formik just ignores it... like Formik set `error` but `touched` is false... also for submit... I must add null or empty strings to initialValues to start showing errors

**Comment by carlosthe19916** [comment]
Using `null` or empty strings also worked for me but I think this should be considered just as a workaround since `undefined` values shouldn't be removed from `values`

**Comment by Kepro** [comment]
also, this is the issue when using `initialValues` and you don't use all values, user click "submit" and we never show error on fields that are emptym because of `undefined`

**Comment by duc-tm** [comment]
Still unresolved in 2023?

**Comment by DarkPage** [comment]
I got the same problem. The onChange function set field value to undefined. And then the form shows no error message. Because the errors and values object has delete the field key.
2023-11-02  I am crazy.

**Comment by sylvaindubus** [comment]
Same problem here, and it took a while to figure out that my issue was this :(

**Comment by tlmader** [comment]
Just ran into this issue myself. Had to work around it by using `setValues`:
```ts
  const handleClear = (name: keyof FormValues) => () => {
    navigation.setParams({
      [name]: undefined,
    });
    /**
     * setFieldValue to undefined deletes the value from values
     * https://github.com/jaredpalmer/formik/issues/2332
     */
    setValues((previous) => ({
      ...previous,
      [name]: undefined,
    }));
  };
  ```

**Comment by souvik1000** [comment]
This issue is still there. On clearing the selection, once we try to set the value as undefined, on submit we are not getting the value with the version of "^2.1.5". So, as a work around, currently I'm overlapping the current outcome with the initialValue. Cause, if some value is not coming, I'm expecting that, it will be undefined only.
```
const initialCustomerFilterValues = {
    customerStatus: undefined,
    organizationStatus: undefined
}

<Formik
    initialValues={initialCustomerFilterValues}
    onSubmit={(values) => {
        filterHandler({ ...initialCustomerFilterValues, ...values })
   }}
>
    ...
</Formik>
```

**Comment by dbnl-nickpalumbo** [comment]
+1 this issue is very frustrating, as it breaks `formik.dirty` which treats `property: undefined` as being different from the property not being present.
